### PR TITLE
Alugrid integration

### PR DESCRIPTION
**Pull request by ElyesAhmed**
This is a first attempt to integrate Dune Alugrid with opm, most changes were  to
(1) add CartesianIndexMapper as an argument for EclTransmissibility to cope with AluCartesianMapper
(2) remove some of the UgGridHelpers
(3) re-ordering cells for computing transmissibility  
(4) Adding Instantiations for  Grid:Alugrid

### CODE REVIEW

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
I haven't looked at the details here yet, but I notice that the branch fails to compile on the Jenkins system.  I **think** that's because ALUGrid is not available there, but maybe @akva2 can confirm.  If the compile failure **is** due to missing ALUGrid support, then we at minimum need to ensure that the branch will compile in the absence of ALUGrid.  That, I think, will require adding `HAVE_ALUGRID` in a few more places such as [`ebos/eclgenericwriter.hh`](https://github.com/OPM/opm-simulators/blob/9d61b59d731b4cbefe9f3e4f1ddbd962ff087032/ebos/eclgenericwriter.hh#L33).

**Comment by akva2** [comment]
yes there is no alugrid available at the moment, providing it will be a bit of work as it's not avail in apt. i'll get working on that but in the mean time you can work on fixing the build without as we definitely need that working.

**Comment by bska** [comment]
jenkins build this please

**Comment by blattms** [comment]
There still seem to be some `#ifdef HAVE_ALUGRID` missing and compilation fails because of this.

File: ebos/eclalugridvanguard.hh [hunk 1]
```diff
@@ -187,8 +239,38 @@ public:
...
+
+    const std::vector<int>& globalCell()
+    {
+        return cartesianCellId_;
+    }
+
+    std::vector<int> cellPartition() const
+    {
+        return cell_part_;
+    }
```
Comment (blattms) [inline ebos/eclalugridvanguard.hh:264]: Does this make sense for ALUGrid? Will it be used?

File: ebos/eclalugridvanguard.hh [hunk 2]
```diff
@@ -200,42 +282,72 @@ protected:
...
-                                         /*flipNormals=*/false,
-                                         /*clipZ=*/false);
+        const EclipseGrid* input_grid = nullptr;
+        std::vector<double> global_porv;
+        // At this stage the ParallelEclipseState instance is still in global
+        // view; on rank 0 we have undistributed data for the entire grid, on
+        // the other ranks the EclipseState is empty.
+        if (mpiRank == 0) {
+            input_grid = &this->eclState().getInputGrid();
+            global_porv = this->eclState().fieldProps().porv(true);
```
Comment (blattms) [inline ebos/eclalugridvanguard.hh:292]: Should we prevent the copy?

File: ebos/eclproblem.hh [hunk 3]
```diff
@@ -867,7 +862,18 @@ public:
...
-        transmissibilities_.finishInit();
+        
+        // Re-ordering in case of Alugrid
+        using Grid = GetPropType<TypeTag, Properties::Grid>;
+        typename std::is_same<Grid, Dune::ALUGrid<3, 3, Dune::cube, Dune::nonconforming>>::type isAlugrid;
+        std::vector<unsigned int> equilGridToGrid; 
+        std::vector<unsigned int> GridToequilGrid;
+        if (isAlugrid) {
+           for (size_t index = 0; index<simulator.vanguard().cartesianSize(); index ++) {
+ equilGridToGrid.push_back(simulator.vanguard().gridEquilIdxToGridIdx(index));       GridToequilGrid.push_back(simulator.vanguard().gridIdxToEquilGridIdx(index));
```
Comment (blattms) [inline ebos/eclproblem.hh:873]: Please fix, indentation and linebreak?

File: ebos/eclproblem.hh [hunk 4]
```diff
@@ -1010,6 +1016,16 @@ public:
...
         const auto& schedule = simulator.vanguard().schedule();
         const auto& events = schedule[episodeIdx].events();
+        
+        // Re-ordering in case of Alugrid
+        using Grid = GetPropType<TypeTag, Properties::Grid>;
+        typename std::is_same<Grid, Dune::ALUGrid<3, 3, Dune::cube, Dune::nonconforming>>::type isAlugrid;
+        std::vector<unsigned int> equilGridToGrid;
+        if (isAlugrid) {
+           for (size_t index = 0; index<simulator.vanguard().cartesianSize(); index ++) {
+    equilGridToGrid.push_back(simulator.vanguard().gridEquilIdxToGridIdx(index));
```
Comment (blattms) [inline ebos/eclproblem.hh:1026]: Please fix indentation.

File: ebos/eclproblem.hh [hunk 5]
```diff
@@ -1161,13 +1177,25 @@ public:
...
         auto& ecl_state = simulator.vanguard().eclState();
         int episodeIdx = this->episodeIndex();
+        // re-ordering in case of ALugrid
+        using Grid = GetPropType<TypeTag, Properties::Grid>;
+        typename std::is_same<Grid, Dune::ALUGrid<3, 3, Dune::cube, Dune::nonconforming, Dune::ALUGridMPIComm>>::type isAlugrid;
+
+        std::vector<unsigned int> GridToequilGrid;
+        if (isAlugrid) {
+           for (size_t index = 0; index<simulator.vanguard().cartesianSize(); index ++) {
+    GridToequilGrid.push_back(simulator.vanguard().gridIdxToEquilGridIdx(index));
```
Comment (blattms) [inline ebos/eclproblem.hh:1187]: Indentation

File: ebos/eclproblem.hh [hunk 6]
```diff
@@ -1279,7 +1307,7 @@ public:
       the simulator properties which need to be updated. This functionality is
       probably not complete.
     */
-    void applySimulatorUpdate(int report_step, Parallel::Communication comm, const SimulatorUpdate& sim_update, EclipseState& ecl_state, Schedule& schedule, SummaryState& summary_state, bool& commit_wellstate) {
+    void applySimulatorUpdate(int report_step, Parallel::Communication comm, const SimulatorUpdate& sim_update, EclipseState& ecl_state, Schedule& schedule, SummaryState& summary_state, bool& commit_wellstate, std::vector<unsigned int> map = {}) {
```
Comment (blattms) [inline ebos/eclproblem.hh:1310]: use `..., const vector& = ...)` to prevent copying?

File: ebos/eclproblem.hh [hunk 7]
```diff
@@ -1302,7 +1330,8 @@ public:
                       EclipseState& ecl_state,
                       Schedule& schedule,
                       Action::State& actionState,
-                      SummaryState& summaryState) {
+                      SummaryState& summaryState,
+                      std::vector<unsigned int> map = {}) {
```
Comment (blattms) [inline ebos/eclproblem.hh:1334]: use `..., const vector& = ...)` to prevent copying?

File: ebos/ecltransmissibility.hh [hunk 8]
```diff
@@ -116,16 +120,16 @@ public:
      * permeabilities. This approach is probably not always correct
      * either but at least it seems to be much better.
      */
-    void finishInit()
-    { update(true); }
+    void finishInit(std::vector<unsigned int> map = {})
```
Comment (blattms) [inline ebos/ecltransmissibility.hh:123]: use `..., const vector& = ...)` to prevent copying?

File: opm/core/props/satfunc/RelpermDiagnostics.cpp [hunk 9]
```diff
@@ -38,6 +38,7 @@
 
 #include <opm/grid/CpGrid.hpp>
 #include <opm/grid/polyhedralgrid.hh>
+#include <dune/alugrid/grid.hh>
```
Comment (blattms) [inline opm/core/props/satfunc/RelpermDiagnostics.cpp:41]: Don't we need an `#ifdef`?

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 10]
```diff
@@ -51,14 +51,17 @@ namespace Opm {
         // Number of cells the global grid view
         global_num_cells_ = ebosSimulator_.vanguard().globalNumCells();
 
-        // Set up parallel wells
-        auto& parallel_wells = ebosSimulator.vanguard().parallelWells();
+        // Set up cartesian mapping.
+        {
+            setupCartesianToCompressed_();
+            auto& parallel_wells = ebosSimulator.vanguard().parallelWells();
```
Comment (blattms) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:57]: Should this be const?

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 11]
```diff
@@ -1517,8 +1521,20 @@ namespace Opm {
...
+        const auto& cartMapper = ebosSimulator_.vanguard().cartesianIndexMapper();
+        const size_t cartesianSize = cartMapper.cartesianSize();
+        cartesian_to_compressed_.resize(cartesianSize, -1);
 
-
+        for (unsigned i = 0; i < local_num_cells_; ++i) {
+            unsigned cartesianCellIdx = cartMapper.cartesianIndex(i);
+            cartesian_to_compressed_[cartesianCellIdx] = i;
+        }
+    }
```
Comment (blattms) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:1537]: Do we maybe have access ti an existing one (ebos/eclbasevanguard.hh, ebos/eclproblem.hh) here and can skip this?

**Review by blattms — CHANGES_REQUESTED** [review]
Did a quick review out of curiosity.
Addtionally to the comments before:
 - Please make sure that we prevent users from misusing this. E.g. If Parallel doe not work with ALUGrid then please detect and throw an expceptions
- Might be useful to add some more tests.

I am curios how the work flow is in parallel: Which grid(s) get loadbalanced? When does ALUGrid get set up and how?

**Comment by bska** [comment]
jenkins build this please

File: ebos/eclgenericwriter.cc [hunk 12]
```diff
@@ -28,6 +28,7 @@
 #include <opm/grid/cpgrid/GridHelpers.hpp>
 #include <opm/grid/polyhedralgrid.hh>
 #include <opm/grid/utility/cartesianToCompressed.hpp>
+#include "eclalugridvanguard.hh"
```
Comment (blattms) [inline ebos/eclgenericwriter.cc:31]: There are still compile errors:
```

In file included from ../../ebos/eclalugridvanguard.hh:32,
                 from ../../ebos/eclgenericwriter.cc:31:
../../ebos/alucartesianindexmapper.hh:33:10: fatal error: dune/alugrid/grid.hh: No such file or directory
   33 | #include <dune/alugrid/grid.hh>
      |          ^~~~~~~~~~~~~~~~~~~~~~
```
Please move this include inside `#if HAVE_DUNE_ALUGRID`, to resolve them.

**Pull request closed without merging**